Support sorting HTML export

libklein opened the pull request:
Resolved #6164.

- Implement sorting support in HtmlExporter
- Add ExportDialog class and UI, which allows to configure export options.

## Screenshots

Current integration:
![kpx1](https://user-images.githubusercontent.com/42714034/135755196-8236fd5f-11a8-4394-8b92-0e184186f626.png)

The export dialog:
![kpx2](https://user-images.githubusercontent.com/42714034/135755190-7ac81281-510f-429e-a1ea-ce33fbc1c751.png)

Export in descending order:
![descending-order](https://user-images.githubusercontent.com/42714034/135755093-3c65e6d5-bd99-457c-b515-5caa97aaa170.png)
Export in ascending order:
![ascending-order](https://user-images.githubusercontent.com/42714034/135755088-b9e95e67-cfb2-4030-bcb7-611a6a9c55f2.png)
Export in database order:
![database-order](https://user-images.githubusercontent.com/42714034/135755091-398c8355-92e2-4d36-94b4-d5c1665c2d70.png)

## Testing strategy
Test on local databases.



## Type of change
- ✅ New feature (change that adds functionality)
- Implements feature request #6164 

## Open issues
- The "ascending" option has no effect when database order is selected. Should the checkbox be disabled in that case or should we remove the "database order" option? I think it probably only causes confusion anyway.
- How about testing? I'd propose to compare the generated .html files to expected output on a test database. I'd reuse the ones in tests/data/*.kdbx for this purpose.

## Discussion / Future work

### Common Exporter Interfance
Both CSV & HTML Exporter share a lot of functionality, especially if we want to include support for sorting, exporting only specific groups, etc. in the future. This should also make implementing new exporters easier. I think we should address this issue and refactor in a future PR. 
- Common Exporter Dialog
- Move Enum to Exporter

libklein commented:
Could someone perhaps give some hints/point me to some information regarding the translation failure?

Also, are there any options on the future work / open issues sections?

droidmonkey commented:
For translation run `./release-tool i18n lupdate`

I'm not sure what you mean by your second question

libklein commented:
Thanks! Did that.

I was talking about the questions mentioned in the "open issues" section of the PR:

> The "ascending" option has no effect when database order is selected. Should the checkbox be disabled in that case or should we remove the "database order" option? I think it probably only causes confusion anyway.

and

> How about testing? I'd propose to compare the generated .html files to expected output on a test database. I'd reuse the ones in tests/data/*.kdbx for this purpose.

droidmonkey commented:
I would remove the Ascending selection box entirely. Give three choices in the "Sort by..." drop down:
* Name Ascending (A-Z)
* Name Descending (Z-A)
* Database Order

For testing, I am not too hung up on test cases for this, it isn't a critical feature. If you want to write them though, use the NewDatabase.kdbx

libklein commented:
The original idea behind the "Ascending" checkbox was that i'd like to extend the system to cover several exporters/be more generic. If output is sortable by several columns, then an ascending checkbox is quickly worth it's weight.

I think this is out of scope for this PR though. I've hence updated the dialog according to your suggestions. 

I have a final question though: How do you properly associate an enum value with a string? Is there some way to leverage the Qt Property/Meta Type system? I've used a QMap for this purpose (with fallback to the enum key). I've tried to wrap the enums key in a call to tr() and insert the proper strings in the translation table, but that does not work with automatic translation generation :/.

droidmonkey commented:
Use a function that maps an enum to a string in a switch statement. 

For extending to this other exporters, like csv, it would be far more useful to have a "sort column" choice in addition to the three sort by choices. Ascending checkbox doesn't make sense on its own.

libklein commented:
Thanks @droidmonkey ! I've updated the code accordingly. It should be ready for merging now.

libklein committed:
Add sorting of HTML export

- Closes #6164
- Implement sorting support in HtmlExporter
- Add ExportDialog class and UI, which allows to configure export options.

libklein commented:
Is there anything else that needs to be done? If not, could you perhaps label the PR as "hacktoberfest-accepted" (see https://hacktoberfest.digitalocean.com/resources/maintainers)? I

droidmonkey commented:
I might do some style changes but otherwise I'll approve the PR for Hacktoberfest reasons

droidmonkey commented:
I made the changes I wanted to make (simplified the code quite a bit) and this is ready for merge.
